va/x11/va_dri3.c

```diff
@@ -0,0 +1,152 @@
+ * limitation the rights to use, copy, modify, merge, publish, distribute,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, provided that the above copyright
+ * notice(s) and this permission notice appear in all copies of the Soft-
+ * ware and that both the above copyright notice(s) and this permission
+ * notice appear in supporting documentation.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABIL-
+ * ITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT OF THIRD PARTY
+ * RIGHTS. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR HOLDERS INCLUDED IN
+ * THIS NOTICE BE LIABLE FOR ANY CLAIM, OR ANY SPECIAL INDIRECT OR CONSE-
+ * QUENTIAL DAMAGES, OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
+ * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER
+ * TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR PERFOR-
+ * MANCE OF THIS SOFTWARE.
+ *
+ * Except as contained in this notice, the name of a copyright holder shall
+ * not be used in advertising or otherwise to promote the sale, use or
+ * other dealings in this Software without prior written authorization of
+ * the copyright holder.
+ *
+ * Authors:
+ *   Emil Velikov (emil.velikov@collabora.com)
+ */
+
+#include "sysdeps.h"
+#include <fcntl.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+
+#include <xcb/xcb.h>
+#include <xcb/dri3.h>
+
+#include <X11/Xlib-xcb.h>
+#include <xf86drm.h>
+
+#include "va_backend.h"
+#include "va_drmcommon.h"
+#include "drm/va_drm_utils.h"
+
+static xcb_screen_t *
+va_DRI3GetXCBScreen(xcb_connection_t *conn, int screen)
+{
+    xcb_screen_iterator_t iter;
+
+    iter = xcb_setup_roots_iterator(xcb_get_setup(conn));
+    for (; iter.rem; --screen, xcb_screen_next(&iter))
+        if (screen == 0)
+            return iter.data;
+    return NULL;
+}
+
+static int
+va_isDRI3Connected(VADriverContextP ctx, int *outfd)
+{
+    xcb_connection_t *conn = XGetXCBConnection(ctx->native_dpy);
+    xcb_screen_t *screen = va_DRI3GetXCBScreen(conn, ctx->x11_screen);
+    xcb_window_t root;
+    const xcb_query_extension_reply_t *ext;
+    xcb_dri3_open_cookie_t cookie;
+    xcb_dri3_open_reply_t *reply;
+    int fd;
+    char *render_node;
+
+    if (!screen)
+        return -1;
+
+    root = screen->root;
+
+    xcb_prefetch_extension_data(conn, &xcb_dri3_id);
+    ext = xcb_get_extension_data(conn, &xcb_dri3_id);
+    if (!ext || !ext->present)
+        return -1;
+
+    /* We don't require any of the ancy stuff, so there's no point in checking
+     * the version.
+     */
+
+    cookie = xcb_dri3_open(conn, root, 0 /* provider */);
+    reply = xcb_dri3_open_reply(conn, cookie, NULL /* error */);
+
+    if (!reply || reply->nfd != 1) {
+        free(reply);
+        return -1;
+    }
+
```

Comment:
On the next line we get fd **from** the XCB. Question is: who owns it now and who is responsible for `close()`?

Gut feeling says that libva now owns it and we must close it on destroy. Destroy path is implemented for wayland (for it's a different - for wayland we open fd in libva, I just note the implementation of destroy):

https://github.com/intel/libva/blob/0a989c9f0af69dcb15253945228c53780f019b79/va/wayland/va_wayland_drm.c#L175

Is the `close()` for fd already implemented somewhere or we need to add?

Reply:
The `close()` is literally a few lines further down.

Bear in mind that the wayland implementation has a series of bugs and memory leaks (as do the X11 code). Assuming yourself and other maintainers have the time to review patches, I can happily open MRs.

Although if the team feels like granting me commit access, I won't object - I will still post my changes as MRs and even help other outstanding MRs.

Reply:
I see that you have added this code to va_x11.c:
```
    if (dri_state->base.fd != -1)
       close(dri_state->base.fd);
```

which closes the fd. That's what I was looking for.